cpp/src/arrow/type.h

```diff
@@ -1015,25 +1018,12 @@ class ARROW_EXPORT Decimal128Type : public DecimalType {
 /// \brief Concrete type class for union data
 class ARROW_EXPORT UnionType : public NestedType {
  public:
-  static constexpr Type::type type_id = Type::UNION;
   static constexpr int8_t kMaxTypeCode = 127;
   static constexpr int kInvalidChildId = -1;
 
-  static constexpr const char* type_name() { return "union"; }
-
-  UnionType(const std::vector<std::shared_ptr<Field>>& fields,
```

Comment:
I'm not sure it's a good idea to remove these constructors, while external code may rely on them.

Reply:
I'm not sure what you mean; UnionType is an abstract base class now and should never be constructed except as a base object of SparseUnionType or DenseUnionType. If we want to maintain compatibility of code which constructs UnionType directly then we'll need to have a single concrete `UnionType` which corresponds to both `Type::SPARSE_UNION` and `Type::DENSE_UNION`. That's possible but very much not our pattern with `DataType` subclasses.

Reply:
Ah, I see. At least keep the static factory function then?